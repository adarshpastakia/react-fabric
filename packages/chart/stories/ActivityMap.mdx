import { Collapsable } from "@react-fabric/core";
import {
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { Fragment } from "react";
import { ActivityMap } from "../src";
import * as ActivityMapStories from "./ActivityMap.stories";

<Meta of={ActivityMapStories} />

<Title>Activity Map</Title>

<Anchor storyId="animations">
  <div className="hidden-anchor">
    <Subheading>Activity Map</Subheading>
  </div>
</Anchor>

<Description of={ActivityMap} />

<Primary />

<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ActivityMapStories.Example} />
</Collapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />

<Anchor storyId="data">
  <Subheading>Data Format</Subheading>
</Anchor>

```ts
// Hours of day
{
    time: "day-hour",
    data:[
        // day 1
        [[hour, day, count],...],
        // day 2
        [[hour, day, count],...]
    ]
}

// Days of month
{
    time: "month-day",
    data:[
        // month 1
        [[day, month, count],...],
        // month 2
        [[day, month, count],...]
    ]
}
```
