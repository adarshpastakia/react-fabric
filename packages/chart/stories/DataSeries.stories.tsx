/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { faker } from "@faker-js/faker";
import { Button, Panel } from "@react-fabric/core";
import type { Meta, StoryObj } from "@storybook/react";
import { useCallback, useEffect, useState } from "react";
import { DataSeries } from "../src";

const meta: Meta<typeof DataSeries> = {
  component: DataSeries,
  title: "@charts/DataSeries",
  parameters: {
    layout: "centered",
    controls: { exclude: "children" },
  },
};

export default meta;
type Story = StoryObj<typeof DataSeries>;

export const Example: Story = {
  render: (args) => {
    const [data, setData] = useState<AnyObject>({});
    const loadData = useCallback(() => {
      const categories = ["Jan", "Feb", "Mar", "Apr", "May", "Jun"];
      setData({
        categoryAxisName: "Months",
        valueAxisName: "Items",
        categories,
        data: [
          {
            id: faker.string.alpha(5),
            label: faker.animal.dog(),
            values: categories.map(() =>
              faker.number.int({ min: 100, max: 500 }),
            ),
          },
          {
            id: faker.string.alpha(5),
            label: faker.animal.dog(),
            values: categories.map(() =>
              faker.number.int({ min: 100, max: 500 }),
            ),
          },
          {
            id: faker.string.alpha(5),
            label: faker.animal.dog(),
            values: categories.map(() =>
              faker.number.int({ min: 100, max: 500 }),
            ),
          },
        ],
      });
    }, []);
    useEffect(() => {
      loadData();
    }, []);
    return (
      <Panel
        width="48rem"
        height="24rem"
        title="Data Series chart"
        expandable
        actions={
          <Button
            aria-label="loadData"
            variant="link"
            icon="mdi mdi-refresh"
            onClick={loadData}
          />
        }
      >
        <DataSeries {...args} {...data} />
      </Panel>
    );
  },
  args: {},
};
