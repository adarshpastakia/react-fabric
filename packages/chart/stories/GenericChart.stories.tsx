/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { Button, Panel, ToggleButtonGroup } from "@react-fabric/core";
import type { Meta, StoryObj } from "@storybook/react";
import React, { useEffect, useState } from "react";
import { GenericChart } from "../src";
import coffeeJson from "./coffee";
import sankeyJson from "./sankey";

const meta: Meta<typeof GenericChart> = {
  component: GenericChart,
  title: "@charts/GenericChart",
  parameters: {
    layout: "centered",
    controls: { exclude: "children" },
  },
};

export default meta;
type Story = StoryObj<typeof GenericChart>;

export const TreeMap: Story = {
  render: (args) => {
    const [type, setType] = useState("treemap");
    const [series, setSeries] = useState<AnyObject>({});

    useEffect(() => {
      if (type === "treemap") {
        setSeries({
          id: "coffee-list",
          name: "Coffee list",
          type: "treemap",
          data: coffeeJson,
          universalTransition: true,
          leafDepth: 2,
          upperLabel: {
            show: true,
            height: 30,
          },
          levels: [
            {
              itemStyle: {
                borderColor: "#777",
                borderWidth: 0,
                gapWidth: 1,
              },
              upperLabel: {
                show: false,
              },
            },
            {
              colorSaturation: [0.35, 0.5],
              itemStyle: {
                borderWidth: 5,
                gapWidth: 1,
                borderColorSaturation: 0.6,
              },
            },
          ],
        });
      }
      if (type === "sunburst") {
        setSeries({
          id: "coffee-list",
          name: "Coffee list",
          type: "sunburst",
          radius: ["20%", "90%"],
          data: coffeeJson,
          universalTransition: true,
          levels: [
            {},
            {
              label: {
                rotate: "tangential",
              },
            },
            {
              r0: "35%",
              r: "70%",
              label: {
                align: "right",
              },
            },
            {
              label: {
                position: "outside",
              },
            },
          ],
        });
      }
    }, [type]);
    return (
      <Panel width="48rem" height="24rem" title="Generic chart" expandable>
        <GenericChart
          {...args}
          series={[series]}
          legend={{ show: false }}
          actions={
            <ToggleButtonGroup value={type} onChange={setType}>
              <Button
                value="treemap"
                icon="mdi mdi-view-dashboard-variant"
                aria-label="tree"
              />
              <Button
                value="sunburst"
                icon="mdi mdi-circle-double"
                aria-label="burst"
              />
            </ToggleButtonGroup>
          }
        />
      </Panel>
    );
  },
  args: {},
};

export const Sankey: Story = {
  render: (args) => {
    const [series, setSeries] = useState<AnyObject>({});

    useEffect(() => {
      setSeries({
        type: "sankey",
        emphasis: {
          focus: "adjacency",
        },
        nodeAlign: "right",
        data: sankeyJson.nodes,
        links: sankeyJson.links,
        lineStyle: {
          color: "source",
          curveness: 0.5,
        },
      });
    }, []);
    return (
      <Panel width="48rem" height="24rem" title="Generic chart" expandable>
        <GenericChart {...args} series={[series]} legend={{ show: false }} />
      </Panel>
    );
  },
  args: {},
};
