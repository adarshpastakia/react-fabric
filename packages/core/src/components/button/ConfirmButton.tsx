/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  FloatingArrow,
  FloatingFocusManager,
  FloatingPortal,
  arrow,
  autoUpdate,
  flip,
  offset,
  shift,
  useClick,
  useDismiss,
  useFloating,
  useInteractions,
} from "@floating-ui/react";
import { mergeRefs } from "@react-fabric/utilities";
import classNames from "classnames";
import {
  Fragment,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from "react";
import { useTranslation } from "react-i18next";
import {
  type CallbackReturn,
  type PolymorphicProps,
  type RefProp,
} from "../../types";
import { Button, type BaseProps, type ButtonProps } from "./Button";

interface ExtendedProps extends Omit<BaseProps, "onClick">, RefProp {
  /**
   * confirm message
   */
  message: string;
  /**
   * ok label
   */
  okLabel?: string;
  /**
   * ok label
   */
  cancelLabel?: string;
  /**
   * click handler, return Promise to autoset loading state
   */
  onClick?: (e: boolean) => CallbackReturn;
}

export type ConfirmButtonProps = ExtendedProps & ButtonProps;

/**
 * Button with confirmation dropdown
 */
export const ConfirmButton = <Tag extends React.ElementType = "button">({
  ref,
  message,
  color,
  okLabel,
  cancelLabel,
  onClick,
  loading,
  actionMessage,
  showActionDone,
  showActionDoneEvent = "click",
  ...props
}: ConfirmButtonProps & PolymorphicProps<Tag>) => {
  const { t } = useTranslation("core");
  const arrowRef = useRef(null);
  const [isOpen, setIsOpen] = useState(false);
  const [busy, setBusy] = useState(false);
  const [actionDone, setActionDone] = useState(false);

  const clickHandler = useCallback(
    (e: boolean) => {
      setBusy(e);
      const ret = onClick?.(e);
      if (e) {
        void Promise.resolve(ret).then((b) => {
          setBusy(false);
          if (b !== false && actionMessage && showActionDoneEvent === "click") {
            setActionDone(true);
          }
        });
      }
    },
    [onClick, actionMessage, showActionDoneEvent],
  );
  useEffect(() => {
    if (actionDone) {
      const timer = setTimeout(() => setActionDone(false), 2000);
      return () => clearTimeout(timer);
    }
  }, [actionDone]);

  const { refs, floatingStyles, context } = useFloating({
    open: isOpen,
    onOpenChange(nextOpen, _, reason) {
      setIsOpen(nextOpen);
      if (reason === "escape-key") {
        clickHandler(false);
      }
    },
    strategy: "fixed",
    placement: "bottom",
    whileElementsMounted: autoUpdate,
    middleware: [
      shift({ padding: 8 }),
      flip(),
      offset(9),
      arrow({
        element: arrowRef,
      }),
    ],
  });

  const click = useClick(context, {
    enabled: !props.disabled,
  });
  const dismiss = useDismiss(context);

  const { getReferenceProps, getFloatingProps } = useInteractions([
    dismiss,
    click,
  ]);

  const ButtonWrapper = useCallback(
    ({ ref, ...props }: AnyObject) => <Button ref={ref} {...props} />,
    [],
  );

  const innerRef = useMemo(
    () => mergeRefs(ref, refs.setReference),
    [ref, refs.setReference],
  );

  return (
    <Fragment>
      <ButtonWrapper
        {...props}
        color={color}
        loading={busy || loading}
        showActionDone={actionDone || showActionDone}
        showActionDoneEvent="manual"
        actionMessage={actionMessage}
        ref={innerRef}
        {...getReferenceProps()}
        data-dropdown-open={isOpen}
      />
      {isOpen && (
        <FloatingPortal>
          <FloatingFocusManager modal context={context}>
            <div
              data-color={color}
              className={classNames(
                "select-none not-italic rounded-capped shadow-lg outline outline-muted bg-base",
              )}
              ref={refs.setFloating}
              style={floatingStyles}
              {...getFloatingProps()}
            >
              <div className="font-medium ">
                <p className="py-2 px-4">{message}</p>
                <div className="flex justify-end py-2 px-4 gap-2">
                  <Button
                    size="sm"
                    variant="link"
                    color={color}
                    onClick={() => [clickHandler(false), setIsOpen(false)]}
                  >
                    {cancelLabel ?? t("action.cancel")}
                  </Button>
                  <Button
                    size="sm"
                    variant="solid"
                    color={color}
                    autoFocus
                    onClick={() => [clickHandler(true), setIsOpen(false)]}
                  >
                    {okLabel ?? t("action.ok")}
                  </Button>
                </div>
              </div>
              <FloatingArrow
                ref={arrowRef}
                context={context}
                strokeWidth={0.5}
                className="fill-base stroke-muted"
              />
            </div>
          </FloatingFocusManager>
        </FloatingPortal>
      )}
    </Fragment>
  );
};
