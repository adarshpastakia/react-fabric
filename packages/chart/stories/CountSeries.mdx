import { Collapsable } from "@react-fabric/core";
import {
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { Fragment } from "react";
import { CountSeries } from "../src";
import * as CountSeriesStories from "./CountSeries.stories";

<Meta of={CountSeriesStories} />

<Title>Count Series</Title>

<Anchor storyId="animations">
  <div className="hidden-anchor">
    <Subheading>Count Series</Subheading>
  </div>
</Anchor>

<Description of={CountSeries} />

<Primary />

<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={CountSeriesStories.Example} />
</Collapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />

<Anchor storyId="data">
  <Subheading>Data Format</Subheading>
</Anchor>

```ts
// Single list of count values
[{ id, label, count },...];
```
