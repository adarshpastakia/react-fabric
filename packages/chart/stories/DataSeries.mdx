import { Collapsable } from "@react-fabric/core";
import {
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { Fragment } from "react";
import { DataSeries } from "../src";
import * as DataSeriesStories from "./DataSeries.stories";

<Meta of={DataSeriesStories} />

<Title>Data Series</Title>

<Anchor storyId="animations">
  <div className="hidden-anchor">
    <Subheading>Data Series</Subheading>
  </div>
</Anchor>

<Description of={DataSeries} />

<Primary />

<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={DataSeriesStories.Example} />
</Collapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />

<Anchor storyId="data">
  <Subheading>Data Format</Subheading>
</Anchor>

```ts
// List of data series
[{ id, label, values: [...] },...];
```
