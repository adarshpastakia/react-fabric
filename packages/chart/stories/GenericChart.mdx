import { Collapsable } from "@react-fabric/core";
import {
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { Fragment } from "react";
import { GenericChart } from "../src";
import * as GenericChartStories from "./GenericChart.stories";

<Meta of={GenericChartStories} />

<Title>Generic Chart</Title>

<Anchor storyId="animations">
  <div className="hidden-anchor">
    <Subheading>Tree Map / Sunburst</Subheading>
  </div>
</Anchor>

<Description of={GenericChart} />

<Primary />

<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={GenericChartStories.TreeMap} />
</Collapsable>

<Anchor storyId="sankey">
  <Subheading>Sankey</Subheading>
</Anchor>
<Description of={GenericChartStories.Sankey} />
<Canvas withToolbar of={GenericChartStories.Sankey} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={GenericChartStories.Sankey} />
</Collapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />
