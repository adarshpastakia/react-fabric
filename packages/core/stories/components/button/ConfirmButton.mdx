import {
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Heading,
  Meta,
  Subheading,
} from "@storybook/blocks";
import { Fragment } from "react";
import { ConfirmButton, Collapsable } from "../../../src";
import * as ConfirmButtonStories from "./ConfirmButton.stories";

<Meta of={ConfirmButtonStories} />

<Title>Confirm Button</Title>

<Anchor storyId="confirmButton">
  <div className="hidden-anchor">
    <Subheading>Confirm Button</Subheading>
  </div>
</Anchor>

<Description of={ConfirmButton} />

<Primary />

<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ConfirmButtonStories._ConfirmButton} />
</Collapsable>

<Anchor storyId="withAction">
  <Subheading>With Action</Subheading>
</Anchor>
<Description of={ConfirmButtonStories.ConfirmWithAction} />
<Canvas withToolbar of={ConfirmButtonStories.ConfirmWithAction} />
<Collapsable className="control-panel">
  <div>Controls</div>
  <Controls of={ConfirmButtonStories.ConfirmWithAction} />
</Collapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" />

<Anchor storyId="sample">
  <Subheading>Example</Subheading>
</Anchor>

```tsx
import { ConfirmButton } from "@react-fabric/core";

export const Example = () => {
  return (
    <ConfirmButton
      icon="icon svg path"
      altIcon="caret icon"
      color="color"
      variant="style"
      loading={boolean}
      disabled={boolean}
      message="Confirmation message"
      okLabel="Yes"
      cancelLabel="No"
    >
      Button Label
    </Button>
  );
};
```
