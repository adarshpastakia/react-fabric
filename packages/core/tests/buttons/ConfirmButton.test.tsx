/*
 * React Fabric
 * @version: 1.0.0
 *
 *
 * The MIT License (MIT)
 * Copyright (c) 2024 Adarsh Pastakia
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy of this software
 * and associated documentation files (the "Software"), to deal in the Software without restriction,
 * including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense,
 * and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all copies or substantial
 * portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
 * TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
 * TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import { composeStories } from "@storybook/react";
import { fireEvent, render, waitFor } from "@testing-library/react";
import * as stories from "../../stories/components/button/ConfirmButton.stories";

const { Tester } = composeStories(stories);

describe("ConfirmButton", () => {
  it("should render confirm button", async () => {
    const fn = jest.fn();
    const fragment = render(
      <Tester message="Confirm?" onClick={fn}>
        Click Me
      </Tester>,
    );
    expect(
      fragment.container.querySelector("[data-ref='button'] > button"),
    ).not.toBeNull();
    expect(
      fragment.container.querySelector("[data-ref='button'] > button > label")
        ?.innerHTML,
    ).toBe("Click Me");

    fireEvent.click(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
    );
    await waitFor(() => {
      expect(
        document.body.querySelector("[data-ref='confirmMessage']")?.innerHTML,
      ).toBe("Confirm?");
      fireEvent.click(
        document.body.querySelector(
          "[data-ref='confirmCancel'] > button",
        ) as HTMLElement,
      );
      expect(fn).toHaveBeenCalledWith(false);
    });
    jest.runAllTimers();

    fireEvent.click(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
    );
    await waitFor(() => {
      expect(
        document.body.querySelector("[data-ref='confirmMessage']")?.innerHTML,
      ).toBe("Confirm?");
      fireEvent.click(
        document.body.querySelector(
          "[data-ref='confirmOk'] > button",
        ) as HTMLElement,
      );
      expect(fn).toHaveBeenCalledWith(true);
    });
    jest.runAllTimers();

    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });

  it("should render confirm button with action", async () => {
    const fragment = render(
      <Tester message="Confirm?" onClick={() => true} actionMessage="Done!">
        Click Me
      </Tester>,
    );
    expect(
      fragment.container.querySelector("[data-ref='button'] > button"),
    ).not.toBeNull();
    expect(
      fragment.container.querySelector("[data-ref='button'] > button > label")
        ?.innerHTML,
    ).toBe("Click Me");

    fireEvent.click(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
    );
    await waitFor(() => {
      expect(
        document.body.querySelector("[data-ref='confirmMessage']")?.innerHTML,
      ).toBe("Confirm?");
      fireEvent.click(
        document.body.querySelector(
          "[data-ref='confirmOk'] > button",
        ) as HTMLElement,
      );
    });
    jest.runAllTimers();

    fireEvent.click(
      fragment.container.querySelector(
        "[data-ref='button'] > button",
      ) as HTMLElement,
    );
    await waitFor(() => {
      expect(
        document.body.querySelector("[data-ref='confirmMessage']")?.innerHTML,
      ).toBe("Confirm?");
      fireEvent.keyDown(
        document.body.querySelector(
          "[data-ref='confirmMessage']",
        ) as HTMLElement,
        {
          key: "Escape",
          code: "Escape",
          shiftKey: false,
          keyCode: 27,
          charCode: 0,
        },
      );
    });
    jest.runAllTimers();

    expect(document.body.innerHTML).toMatchSnapshot();
    fragment.unmount();
  });
});
