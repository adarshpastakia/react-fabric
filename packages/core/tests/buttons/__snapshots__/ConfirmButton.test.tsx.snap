// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ConfirmButton should render confirm button 1`] = `"<div tabindex="0"><div style="display: contents;"><div class="flex gap-2 items-center p-4"><div class="fabric-buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="fabric-button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary" data-dropdown-open="false"><label class="fabric-buttonLabel truncate" data-colored-icon="false">Click Me</label></button></div></div></div></div><div style="z-index: var(--z-overlay);" class="fabric-container__overlays pointer-events-none fixed inset-0 p-4 bg-transparent"></div><div style="z-index: var(--z-notifications);" class="fabric-container__notifs flex flex-nowrap pointer-events-none fixed inset-0 p-2 bg-transparent flex-col-reverse items-center justify-end"></div><div style="z-index: var(--z-notifications);" class="fabric-container__toasts flex flex-nowrap pointer-events-none fixed inset-0 p-4 bg-transparent flex-col items-end justify-end overflow-auto scroll-hide"></div>"`;

exports[`ConfirmButton should render confirm button with action 1`] = `"<div tabindex="0"></div><div tabindex="0"><div style="display: contents;"><div class="flex gap-2 items-center p-4"><div class="fabric-buttonWrapper relative align-text-top select-none focus-within:z-1 inline-block" data-ref="button"><button role="button" class="fabric-button pointer-events-auto cursor-pointer select-none flex flex-nowrap w-full h-full items-center border-0 -outline-offset-1 justify-center max-w-72 p-0 font-sans font-medium appearance-none focus-visible:ring-2 focus-visible:ring-offset-2 buttonInner" type="button" tabindex="0" data-inner-clickable="true" data-color="primary" data-dropdown-open="false"><label class="fabric-buttonLabel truncate" data-colored-icon="false">Click Me</label></button></div></div></div></div><div style="z-index: var(--z-overlay);" class="fabric-container__overlays pointer-events-none fixed inset-0 p-4 bg-transparent"></div><div style="z-index: var(--z-notifications);" class="fabric-container__notifs flex flex-nowrap pointer-events-none fixed inset-0 p-2 bg-transparent flex-col-reverse items-center justify-end"></div><div style="z-index: var(--z-notifications);" class="fabric-container__toasts flex flex-nowrap pointer-events-none fixed inset-0 p-4 bg-transparent flex-col items-end justify-end overflow-auto scroll-hide"></div>"`;
